File: strategy/generator/signal/zerocross/di_zerocross.py
from dataclasses import dataclass

from core.models.parameter import (
    Parameter,
    RandomParameter,
    StaticParameter,
)
from core.models.smooth import Smooth
from core.models.source import SourceType
from strategy.generator.signal.base import Signal, SignalType


@dataclass(frozen=True)
class DiZeroCrossSignal(Signal):
    type: SignalType = SignalType.DiZeroCross
    source_type: Parameter = StaticParameter(SourceType.CLOSE)
    smooth_type: Parameter = StaticParameter(Smooth.WMA)
    period: Parameter = RandomParameter(10.0, 15.0, 1.0)

File: ta_lib/strategies/signal/src/zerocross/di_zerocross.rs
use base::prelude::*;
use core::prelude::*;
use momentum::di;

pub struct DiZeroCrossSignal {
    source_type: SourceType,
    smooth_type: Smooth,
    period: usize,
}

impl DiZeroCrossSignal {
    pub fn new(source_type: SourceType, smooth_type: Smooth, period: f32) -> Self {
        Self {
            source_type,
            smooth_type,
            period: period as usize,
        }
    }
}

impl Signal for DiZeroCrossSignal {
    fn lookback(&self) -> usize {
        self.period
    }

    fn generate(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let di = di(
            &data.source(self.source_type),
            self.smooth_type,
            self.period,
        );

        (di.cross_over(&ZERO_LINE), di.cross_under(&ZERO_LINE))
    }
}
